kv/util/codec/codec.go

```diff
@@ -0,0 +1,81 @@
+package codec
+
+import "github.com/pingcap/errors"
+
+const (
+	encGroupSize = 8
+	encMarker    = byte(0xFF)
+	encPad       = byte(0x0)
+)
+
+var pads = make([]byte, encGroupSize)
+
+// EncodeBytes guarantees the encoded value is in ascending order for comparison,
+// encoding with the following rule:
+//  [group1][marker1]...[groupN][markerN]
+//  group is 8 bytes slice which is padding with 0.
+//  marker is `0xFF - padding 0 count`
+// For example:
+//   [] -> [0, 0, 0, 0, 0, 0, 0, 0, 247]
+//   [1, 2, 3] -> [1, 2, 3, 0, 0, 0, 0, 0, 250]
+//   [1, 2, 3, 0] -> [1, 2, 3, 0, 0, 0, 0, 0, 251]
+//   [1, 2, 3, 4, 5, 6, 7, 8] -> [1, 2, 3, 4, 5, 6, 7, 8, 255, 0, 0, 0, 0, 0, 0, 0, 0, 247]
+// Refer: https://github.com/facebook/mysql-5.6/wiki/MyRocks-record-format#memcomparable-format
+func EncodeBytes(data []byte) []byte {
+	// Allocate more space to avoid unnecessary slice growing.
+	// Assume that the byte slice size is about `(len(data) / encGroupSize + 1) * (encGroupSize + 1)` bytes,
+	// that is `(len(data) / 8 + 1) * 9` in our implement.
+	dLen := len(data)
+	result := make([]byte, 0, (dLen/encGroupSize+1)*(encGroupSize+1)+8) // make extra room for appending ts
+	for idx := 0; idx <= dLen; idx += encGroupSize {
+		remain := dLen - idx
+		padCount := 0
+		if remain >= encGroupSize {
+			result = append(result, data[idx:idx+encGroupSize]...)
+		} else {
+			padCount = encGroupSize - remain
+			result = append(result, data[idx:]...)
+			result = append(result, pads[:padCount]...)
+		}
+
+		marker := encMarker - byte(padCount)
+		result = append(result, marker)
+	}
+	return result
+}
+
+// DecodeBytes decodes bytes which is encoded by EncodeBytes before,
+// returns the leftover bytes and decoded value if no error.
+func DecodeBytes(b []byte) ([]byte, []byte, error) {
```

Comment:
Can we panic here instead of returning the error? I think we do not need proper error handling for TinyKV.

Reply:
already panic in `DecodeKey`, but we need to know whether the key is encoded or raw by checking the error != nil

Reply:
Yeah, I mean it would be nicer to panic in DecodeBytes than DecodeKey. We can return nil, nil if there is an error

Reply:
https://github.com/pingcap-incubator/tinykv/pull/52/files#diff-ee22b14637a12c92c212438b82fb4600R46
See, if we panic in `DecodeBytes`, splitting on a raw key will cause panic cause it doesn't follow the format!

Reply:
It is a bit sad that we cannot always know whether a key is raw or encoded. I suppose the problem is that the client can use the raw API to write raw keys to the DB? But this makes me wonder that one could write a raw key which can be validly decoded, but isn't in fact an encoded key and then things would break. Maybe we should not offer the raw API in TinyKV at all? Does TinySQL use it?

Reply:
If it's a raw key which can be validly decoded, decode and then encode on the key doesn't make any difference, so it won't break the correctness for split checker. 

Also, TinyKV use raw API for LAB1-3 which is to build a kvserver without transaction support